app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt

```diff
@@ -120,10 +126,24 @@ class MaskLayer(private val imageMatrix: Matrix) {
         }
     }
 
+    fun setMaskBitmap(bitmap: Bitmap){
+        require(sizeInitialized()) { "Size was not initialized yet" }
+        initialBitmap = bitmap
+        currentStateBitmap = Bitmap.createScaledBitmap(bitmap, size.width, size.height, false)
+        currentStateBitmapCanvas = Canvas(currentStateBitmap)
+        draw(currentStateBitmapCanvas)
+    }
+
     private fun pathBeingDrawn() = currentPath != null
 
     private fun composeCurrentStateBitmap() {
         currentStateBitmap.eraseColor(Color.TRANSPARENT)
+
+        initialBitmap?.let{
+            currentStateBitmap = Bitmap.createScaledBitmap(it, size.width, size.height, false)
+            currentStateBitmapCanvas = Canvas(currentStateBitmap)
+        }
```

Comment:
Initially I thought that my phone is being slow, but then I realised that it's this that is actually slowing it down. 

This method is ran on each frame, and allocating a bitmap on each frame is really not good for performance. I think that we need to find another way to draw the existing mask that is being edited. I have an idea, but it would involve changing the undo/redo mechanism... Let's catch-up on this topic, if you want.

Reply:
Updated as discussed. Thanks!